crates/shared/src/price_estimation/quasimodo.rs

```diff
@@ -0,0 +1,346 @@
+        let pairs = self.base_tokens.relevant_pairs([token_pair].into_iter());
+
+        let amms = self
+            .pools
+            .fetch(pairs, Block::Recent)
+            .await?
+            .iter()
+            .map(|pool| AmmModel {
+                parameters: AmmParameters::ConstantProduct(ConstantProductPoolParameters {
+                    reserves: BTreeMap::from([
+                        (pool.tokens.get().0, pool.reserves.0.into()),
+                        (pool.tokens.get().1, pool.reserves.1.into()),
+                    ]),
+                }),
+                fee: BigRational::from((
+                    BigInt::from(*pool.fee.numer()),
+                    BigInt::from(*pool.fee.denom()),
+                )),
+                cost: CostModel {
+                    amount: U256::from(GAS_PER_UNISWAP) * gas_cost,
+                    token: self.native_token,
+                },
+                mandatory: false,
+            })
+            .enumerate()
+            .collect();
+
+        let settlement = self
+            .api
+            .solve(
+                &BatchAuctionModel {
+                    tokens,
+                    orders,
+                    amms,
+                    metadata: None,
+                },
+                Instant::now() + Duration::from_secs(1),
+            )
+            .await?;
+
+        if settlement.orders.is_empty() {
+            return Err(PriceEstimationError::NoLiquidity);
+        }
+
+        let mut cost = self.extract_cost(&settlement.orders[&0].cost)?;
+        for amm in settlement.amms.values() {
+            cost += self.extract_cost(&amm.cost)? * amm.execution.len();
+        }
+        let gas = (cost / gas_cost)
+            + INITIALIZATION_COST // Call into contract
+            + SETTLEMENT // overhead for entering the `settle()` function
+            + ERC20_TRANSFER * 2; // transfer in and transfer out
+
+        Ok(Estimate {
+            out_amount: match query.kind {
+                OrderKind::Buy => settlement.orders[&0].exec_sell_amount,
+                OrderKind::Sell => settlement.orders[&0].exec_buy_amount,
+            },
+            gas,
+        })
+    }
+
+    fn extract_cost(&self, cost: &Option<CostModel>) -> Result<U256, PriceEstimationError> {
+        if let Some(cost) = cost {
+            if cost.token != self.native_token {
+                Err(anyhow::anyhow!("cost specified as an unknown token {}", cost.token).into())
+            } else {
+                Ok(cost.amount)
+            }
+        } else {
+            Ok(U256::zero())
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl PriceEstimating for QuasimodoPriceEstimator {
+    async fn estimates(
+        &self,
+        queries: &[Query],
+    ) -> Vec<anyhow::Result<Estimate, PriceEstimationError>> {
+        let mut results = Vec::with_capacity(queries.len());
+
+        for query in queries {
+            results.push(self.estimate(query).await);
+        }
```

Comment:
For a future PR it would be nice to move some shared computation here from the estimate function. Just like for the Baseline price estimator we can precalculate the gas price, token infos, pools and pass them in to avoid duplicate work.